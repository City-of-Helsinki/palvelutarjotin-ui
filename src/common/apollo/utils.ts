import { NormalizedCacheObject, ApolloClient } from '@apollo/client';

export class MutableReference<Ref = unknown> {
  _value?: Ref;

  constructor(initialReference?: Ref) {
    this._value = initialReference;
  }

  get reference(): Ref | undefined {
    return this._value;
  }

  set reference(ref: Ref | undefined) {
    this._value = ref;
  }
}

type InitApolloClientConfig<
  TCacheShape,
  Client extends ApolloClient<TCacheShape>,
> = {
  initialState: TCacheShape;
  mutableCachedClient: MutableReference<Client>;
  createClient: () => Client;
};

// https://www.apollographql.com/blog/building-a-next-js-app-with-apollo-client-slash-graphql/
export function initializeApolloClient<
  TCacheShape = NormalizedCacheObject,
  Client extends ApolloClient<TCacheShape> = ApolloClient<TCacheShape>,
>({
  initialState,
  mutableCachedClient,
  createClient,
}: InitApolloClientConfig<TCacheShape, Client>): ApolloClient<TCacheShape> {
  const _apolloClient = mutableCachedClient.reference ?? createClient();

  // Initial state hydration
  if (initialState) {
    // Get existing cache, loaded during client side data fetching
    const existingCache = _apolloClient.extract();

    // Restore the cache using the data passed from
    // getStaticProps/getServerSideProps combined with the existing cached data
    _apolloClient.cache.restore({ ...existingCache, ...initialState });
  }

  // For SSG and SSR always create a new Apollo Client
  if (typeof window === 'undefined') {
    return _apolloClient;
  }

  // Create the Apollo Client once in the client
  if (!mutableCachedClient.reference) {
    mutableCachedClient.reference = _apolloClient;
  }

  return _apolloClient;
}
