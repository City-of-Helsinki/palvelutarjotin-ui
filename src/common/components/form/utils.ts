/* eslint-disable @typescript-eslint/no-explicit-any */
import { FormikErrors, FormikTouched } from 'formik';
import { TFunction } from 'i18next';

/** Get error text
 * @param {Object} errors
 * @param {Object} touched
 * @param {string} name
 * @param {Function} t
 * @return {string}
 */
export const getErrorText = (
  errors: FormikErrors<any>,
  touched: FormikTouched<any>,
  name: string,
  t: TFunction
): string => {
  const error: any = errors[name];

  return !!error && touched[name]
    ? typeof error === 'string'
      ? t(error)
      : t(error.key, error)
    : '';
};
