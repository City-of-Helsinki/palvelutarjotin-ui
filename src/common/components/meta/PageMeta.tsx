import Head from 'next/head';
import { useRouter } from 'next/router';
import React from 'react';

import { Language } from '../../../types';

export type Props = {
  // Title of page, required for accessibility: pages should have unique titles
  // so that screen reader users are able to determine when the current page is
  // changed.
  title: string;
  description?: string | null;
  image?: string | null;
  canonicalUrl?: string | null;
  openGraphDescription?: string | null;
  openGraphTitle?: string | null;
  openGraphType?: string | null;
  twitterDescription?: string | null;
  twitterTitle?: string | null;
  localePaths?: {
    path: string;
    locale: Language;
  }[];
};

export function getLanguageAwarePath(
  locale: Language | undefined,
  path: string
): string {
  if (locale === 'fi') {
    return path;
  }

  return `/${locale}${path}`;
}

const PageMeta: React.FC<Props> = ({
  title,
  description,
  image,
  openGraphType,
  twitterDescription,
  twitterTitle,
  localePaths,
  ...seo
}) => {
  const { locale, asPath } = useRouter();
  const canonical = getLanguageAwarePath(locale as Language, asPath);
  const openGraphTitle = seo.openGraphTitle ?? title;
  const openGraphDescription = seo.openGraphDescription ?? description;

  return (
    <>
      <Head>
        <title>{title}</title>
        {description && <meta name="description" content={description} />}
        <meta property="og:title" content={openGraphTitle} />
        {openGraphDescription && (
          <meta property="og:description" content={openGraphDescription} />
        )}
        {image && <meta property="og:image" content={image} />}
        {openGraphType && <meta property="og:type" content={openGraphType} />}
        {twitterTitle && <meta name="twitter:title" content={twitterTitle} />}
        {twitterDescription && (
          <meta name="twitter:description" content={twitterDescription} />
        )}
        {localePaths?.map(({ locale, path }) => (
          <link key={locale} rel="alternate" hrefLang={locale} href={path} />
        ))}
        <link rel="canonical" href={canonical} />
        <link
          rel="icon"
          href="/hds-favicon-kit/favicon-32x32.ico"
          sizes="any"
        />
        <link
          rel="icon"
          href="/hds-favicon-kit/favicon.svg"
          type="image/svg+xml"
        />
        <link
          rel="apple-touch-icon"
          href="/hds-favicon-kit/apple-touch-icon.png"
        />
      </Head>
    </>
  );
};

export default PageMeta;
