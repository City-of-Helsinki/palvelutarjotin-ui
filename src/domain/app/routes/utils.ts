import { CollectionItemType } from 'react-helsinki-headless-cms';

import { ROUTES } from './constants';

export const getCmsPagePath = (uri?: string | null): string => {
  return ROUTES.CMS_PAGE.replace('/:uri', uri ?? '');
};

export const getCmsArticlePath = (uri?: string | null): string => {
  return ROUTES.CMS_ARTICLE.replace('/:uri', uri ?? '');
};

export const getCollectionItemUrl = (item: CollectionItemType): string => {
  if (!item) {
    return '#';
  }
  if (item.__typename === 'Post') {
    return getCmsArticlePath(item.uri);
  }
  if (item.__typename === 'Page') {
    return getCmsPagePath(item.uri);
  }
  return item?.uri ?? '';
};
