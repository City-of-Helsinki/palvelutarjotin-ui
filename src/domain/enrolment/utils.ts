import {
  EnrolOccurrenceMutationInput,
  Language,
  NotificationType,
} from '../../generated/graphql';
import { EnrolmentFormFields } from './enrolmentForm/constants';

const getNotificationType = (values: EnrolmentFormFields): NotificationType => {
  return values.hasEmailNotification && values.hasSmsNotification
    ? NotificationType.EmailSms
    : NotificationType.Email;
};

export const getEnrolmentPayload = ({
  occurrenceIds,
  values,
}: {
  occurrenceIds: string[];
  values: EnrolmentFormFields;
}): EnrolOccurrenceMutationInput => {
  return {
    occurrenceIds,
    notificationType: getNotificationType(values),
    person: values.isSameResponsiblePerson
      ? undefined
      : { ...values.person, language: values.language as Language },
    studyGroup: {
      ...values.studyGroup,
      amountOfAdult: Number(values.studyGroup.amountOfAdult),
      groupSize: Number(values.studyGroup.groupSize),
      person: {
        ...values.studyGroup.person,
        language: values.language as Language,
      },
      studyLevels: values.studyGroup.studyLevels,
    },
  };
};

declare let grecaptcha: ReCAPTCHA;

interface ReCAPTCHA {
  ready: (f: () => void) => void;
  execute: (key: string, action: { action: string }) => Promise<string>;
}

export const getCAPTCHAToken = (): Promise<string> => {
  return new Promise((resolve, reject) => {
    if (process.env.NEXT_PUBLIC_CAPTCHA_KEY) {
      const captchaKey = process.env.NEXT_PUBLIC_CAPTCHA_KEY;
      grecaptcha.ready(() => {
        grecaptcha
          .execute(captchaKey, { action: 'submit' })
          .then(resolve)
          .catch(reject);
      });
    } else {
      reject('CAPTCHA_KEY missing!');
    }
  });
};
