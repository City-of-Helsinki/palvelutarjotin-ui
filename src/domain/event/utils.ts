import { isToday, isTomorrow } from 'date-fns';
import { TFunction } from 'next-i18next';

import { EventFieldsFragment } from '../../generated/graphql';
import { Language } from '../../types';
import formatDate from '../../utils/formatDate';
import getLocalisedString from '../../utils/getLocalisedString';
import getTimeFormat from '../../utils/getTimeFormat';
import { EVENT_PLACEHOLDER_IMAGES } from './constants';

/**
 * Get event placeholder image url
 * @param {string} id
 * @return {string}
 */
export const getEventPlaceholderImage = (id: string): string => {
  const numbers = id.match(/\d+/g);
  const sum = numbers
    ? numbers.reduce((prev: number, cur: string) => prev + Number(cur), 0)
    : 0;
  const index = sum % 4;

  return EVENT_PLACEHOLDER_IMAGES[index];
};

export const getEventStartTimeStr = (
  event: EventFieldsFragment,
  locale: Language,
  t: TFunction
): string | null => {
  const startTime = event.startTime ? new Date(event.startTime) : null;
  const timeFormat = getTimeFormat(locale);

  if (!startTime) return null;

  if (isToday(startTime))
    return t('event:eventCard.startTime.today', {
      time: formatDate(startTime, timeFormat, locale),
    });

  if (isTomorrow(startTime))
    return t('event:eventCard.startTime.tomorrow', {
      time: formatDate(startTime, timeFormat, locale),
    });

  return t('event:eventCard.startTime.other', {
    date: formatDate(startTime),
    time: formatDate(startTime, timeFormat, locale),
  });
};

/**
 * Check is event free
 * @param eventData
 * @return {boolean}
 */
export const isEventFree = (event: EventFieldsFragment): boolean =>
  Boolean(event.offers.find((item) => item.isFree)?.isFree);

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types
export const getEventFields = (
  event: EventFieldsFragment | undefined | null,
  locale: Language
) =>
  event
    ? {
        isEventFree: isEventFree(event),
        eventName: getLocalisedString(event?.name, locale),
        shortDescription: getLocalisedString(event?.shortDescription, locale),
        description: getLocalisedString(event?.description, locale),
        imageUrl: event?.images?.[0]?.url,
        imageAltText: event?.images?.[0]?.altText,
        photographerName: event?.images?.[0]?.photographerName,
        organization: event.pEvent?.organisation?.name,
        contactPhoneNumber: event.pEvent?.contactPhoneNumber,
        contactEmail: event.pEvent?.contactEmail,
        contactPerson: event.pEvent?.contactPerson?.name,
      }
    : {};
