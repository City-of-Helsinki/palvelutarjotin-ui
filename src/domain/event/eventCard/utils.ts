import { TFunction } from 'next-i18next';

import { EnrolmentStatusKeywordPropsType } from '../../../common/components/keyword/EnrolmentStatusKeyword';
import { EventsFieldsFragment } from '../../../generated/graphql';
import { Language } from '../../../types';
import { formatDateRange } from '../../../utils/time/format';
import { isMultidayOccurrence } from '../../occurrence/utils';
import { getEventStartTimeStr } from '../utils';

export const getOccurrenceDateStr = (
  occurrence: { startTime: string; endTime: string },
  locale: Language,
  t: TFunction
): string | null => {
  if (isMultidayOccurrence(occurrence)) {
    return formatDateRange(
      new Date(occurrence.startTime),
      new Date(occurrence.endTime)
    );
  }

  return getEventStartTimeStr(new Date(occurrence.startTime), locale, t);
};

/**
 * Has the event enrolments enabled in any of it's occurrences?
 * @param event The event to check the status for.
 * @returns `true` if any of the event occurrences has enrolments and
 *  `false` if none.
 */
export const hasEventEnrolmentsEnabled = (event: EventsFieldsFragment) =>
  Boolean(event.pEvent.enrolmentStart) || hasEventExternalEnrolments(event);

/**
 * Has the event external enrolments enabled in any of it's occurrences?
 *
 * @param event The event to check the status for.
 * @returns `true` if the event has some external enrolments and
 *  `false` if none of the occurrences has external enrolments.
 */
export const hasEventExternalEnrolments = (event: EventsFieldsFragment) =>
  Boolean(event.pEvent.externalEnrolmentUrl);

/**
 * Resolve the enrolment status of an event. Has the event free spots available?
 *
 * @param event The event to check the status for.
 * @returns Returns `null` if the enrolment status could not be resolved or
 *    the event has no enrolments at all,
 *    `true` if there are any free spots available and `false` if there are no
 *    spots available.
 */
export const hasEventFreeSpotsAvailable = (
  event: EventsFieldsFragment
): boolean | null => event.pEvent.hasSpaceForEnrolments ?? null;

/**
 * Determines the enrollment status of an event.
 *
 * @param event The event to check the enrollment status for.
 * @returns The enrollment status of the event, or `undefined`
 *      if no enrolments exist or the status could not be determined.
 */
export const getEventEnrolmentStatus = (
  event: EventsFieldsFragment
): EnrolmentStatusKeywordPropsType['enrolmentStatus'] => {
  if (hasEventEnrolmentsEnabled(event)) {
    if (hasEventExternalEnrolments(event)) {
      return 'externally_unknown';
    }
    const hasEventFreeSpots = hasEventFreeSpotsAvailable(event);

    if (hasEventFreeSpots === false) {
      // Event has no free spots available, so it is full.
      return 'full';
    } else if (hasEventFreeSpots === true) {
      // Event has some free spots available.
      return 'free_space_available';
    }
  }
  // No enrolments at all or the enrolment status could not be resolved
  return undefined;
};
