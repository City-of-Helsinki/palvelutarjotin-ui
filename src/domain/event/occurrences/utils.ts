import {
  EventFieldsFragment,
  OccurrenceFieldsFragment,
} from '../../../generated/graphql';
import { ENROLMENT_ERRORS } from '../../enrolment/constants';
import {
  hasOccurrenceSpace,
  isEnrolmentClosed,
  isOccurrenceCancelled,
} from '../../occurrence/utils';

export function getEnrolmentError(
  occurrence: OccurrenceFieldsFragment,
  event: EventFieldsFragment
): ENROLMENT_ERRORS | null {
  if (isEnrolmentClosed(occurrence, event))
    return ENROLMENT_ERRORS.ENROLMENT_CLOSED_ERROR;
  if (!hasOccurrenceSpace(occurrence))
    return ENROLMENT_ERRORS.NOT_ENOUGH_CAPACITY_ERROR;
  if (isOccurrenceCancelled(occurrence)) {
    return ENROLMENT_ERRORS.ENROLMENT_CANCDELLED_ERROR;
  }

  return null;
}

export function getOrderedLanguages(
  languages: { id: string; name: string }[]
): string[] {
  const languageOrder = ['fi', 'sv', 'en'];
  const languageStrings = languages.map((l) => l.id.split('_')[0]);

  // order languages based on languageOrder, other languages go
  // to the end of the array
  return languageStrings.sort((a, b) => {
    const aIndex = languageOrder.indexOf(a);
    const bIndex = languageOrder.indexOf(b);

    if (aIndex > -1 && bIndex > -1) {
      return aIndex - bIndex;
    }

    if (aIndex > -1 && bIndex === -1) {
      return -1;
    }

    if (bIndex > -1 && aIndex === -1) {
      return 1;
    }

    return 0;
  });
}
