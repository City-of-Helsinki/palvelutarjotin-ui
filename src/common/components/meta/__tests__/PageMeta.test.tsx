import React from 'react';

import { render } from '../../../../utils/testUtils';
import PageMeta from '../PageMeta';

jest.mock('next/head', () => {
  return {
    __esModule: true,
    default: ({ children }: { children: Array<React.ReactElement> }) => {
      return <>{children}</>;
    },
  };
});

// just a simple snapshot test to make sure component is rendering stuff correctly
it('renders meta elements', async () => {
  const { container } = render(
    <PageMeta
      title="Title"
      description="description"
      openGraphDescription="openGraphDescription"
      openGraphTitle="openGraphTitle"
    />
  );

  expect(container).toMatchSnapshot();
});
