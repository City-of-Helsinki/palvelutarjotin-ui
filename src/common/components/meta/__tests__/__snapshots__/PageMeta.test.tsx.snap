// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders meta elements 1`] = `
<div>
  <title>
    Title
  </title>
  <meta
    content="description"
    name="description"
  />
  <meta
    content="openGraphTitle"
    property="og:title"
  />
  <meta
    content="openGraphDescription"
    property="og:description"
  />
  <link
    href="/undefined/"
    rel="canonical"
  />
  <link
    href="/hds-favicon-kit/favicon-32x32.ico"
    rel="icon"
    sizes="any"
  />
  <link
    href="/hds-favicon-kit/favicon.svg"
    rel="icon"
    type="image/svg+xml"
  />
  <link
    href="/hds-favicon-kit/apple-touch-icon.png"
    rel="apple-touch-icon"
  />
</div>
`;
